File: org.glite.wms-ui.api-cpp/src/Shadow.cpp
/* ********************************************************************
* File name: Shadow.cpp
*  author    : Alessandro Maraschini <alessandro.maraschini@datamat.it>
*  copyright : (C) 2002 by DATAMAT
*********************************************************************/
#include "glite/wmsui/api/Shadow.h"
#include "glite/wmsui/api/JobExceptions.h"
#include "glite/jdl/Ad.h"

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#ifdef HAVE_STRINGSTREAM
#include <sstream>
#else
#include <strstream>
#endif

using namespace glite::wmsutils::jobid ; //JobId
using namespace std ;

namespace glite {
namespace wmsui {
namespace api {

	Shadow::Shadow (JobId jid , Listener *ls) {set(jid , ls);}
	Shadow::Shadow (){};
	Shadow::~Shadow(){};
	void Shadow::set(JobId jid ,   Listener* ls) {  jobId = jid ; listener = ls ; } ;
	void Shadow::attach(int port){
		// Query to LB: host & port
		// if Query--> Error :  Do nothing, proceed anyway
		// If old Shadow is up  ---> kill process
		// start the new process (the named pipeline are the same) (if error -> EXC)
		console (port) ;
		// Log new host / port to LB (if error -> EXC)
		// Run listener
	} ;

	/**  Read from the specified named pipe and return the string */
	string Shadow::empty(string bufferName){
		string result="" ,outMsg;
		filebuf fb;
		fb.open (bufferName.c_str() , ios::in);
		istream jdl_in(&fb);
#ifdef HAVE_STRINGSTREAM
		stringstream buffer ;
#else
		strstream buffer;
#endif
		int leng = 2048;
		char tmp[leng];
		while (  jdl_in.getline(tmp,leng,'\n')  ){
		result +="\n" + string(tmp) ;
		}
		fb.close() ;
		return result ;
	} ;
	string Shadow::getPipeErr(){        return pipeRoot +".err" ;    };
	string Shadow::getPipeIn(){       return pipeRoot +".in" ;    };
	string Shadow::getPipeOut(){       return pipeRoot +".out" ;    };
	string Shadow::getHost(){       return "localHost" ; /*TBD ! ! */   };
	int Shadow::getPid(){       return pid ;    };
	int Shadow::getPort(){       return port ;    };
	/*********************
	* kill
	*********************/
	void Shadow::kill (){
		if (pid != 0 ){
			char tmp [2048];
			sprintf(tmp , "%d", pid) ;
			string command = "kill -9 " + string(tmp);
			system(command.c_str());
		}
	}
	/*********************
	* Detach
	*********************/
	void Shadow::detach(){
		kill() ;
		remove(   getPipeErr().c_str()    );
		remove(   getPipeIn().c_str()    );
		remove(   getPipeOut().c_str()    );
	};
	/*********************
	* Start
	*********************/
	void Shadow::start(){
		listener->run( this) ;
	};

	/** Launch the console shadow and read pid and port from the named written pipe   */
	void Shadow::console( int port){
		GLITE_STACK_TRY("Shadow::console()") ;
		// Kill the grid-console-shadow, if still running:
		kill () ;
		pipeRoot = "/tmp/listener-" + jobId.getUnique();
		// Launch the grid-console-shadow with --logfile (background process)
		string arguments = " -log-to-file" +pipeRoot ;
		int fromP =0, toP =0 ;
		char* gl_env  =getenv ("GLOBUS_TCP_PORT_RANGE") ;
		if (gl_env!= NULL){
			// environment variable found
			string gtp_env = string( gl_env  );
			string sep[] = {" " , ":" , "-" , "," } ;
			for (unsigned int i = 0 ; i< sep->length() ; i++){
				//cout << "Looking for '" << sep[i] <<"' in " << gtp_env << flush ;
				if ( gtp_env.find( sep[i] ) <  gtp_env.size()  ){
					fromP = atoi (  gtp_env.substr(0 , gtp_env.find( sep[i] ) ).c_str()   ) ;
					toP    = atoi (  gtp_env.substr(gtp_env.find( sep[i] )  ).c_str()   ) ;
						break;
				}
			}
		}
		if (port !=0){
			if (    (port < fromP  ) || (port > toP)    )
				throw JobOperationException  ( __FILE__ , __LINE__ ,METHOD , WMS_JOBOP_ALLOWED , "Unable to perform attachement: port exceeds firewall range"  ) ;
			arguments += "-port " ; //TBD + repr (port)
		}
		char* shPath_ch = getenv ("GLITE_WMS_LOCATION") ;
		string shPath  ;
		if (  shPath == NULL)
			shPath = "/opt/glite" ;
		shPath = string(shPath_ch) + "/bin/glite-wms-grid-console-shadow"  ;
		string command = shPath  +arguments +" &";
		if (  system(command.c_str())   )
			throw JobOperationException  ( __FILE__ , __LINE__ ,METHOD , WMS_JOBOP_ALLOWED , "Unable to launch the Shadow listaner executable: "+ shPath  ) ;
		// get the pid and the port:
		string adStr ;
		int timeout= 0 ;
		while (timeout<10){
			try{
				adStr += empty( pipeRoot   );
			}catch (exception &exc){
				sleep(1) ;
				timeout++ ;
			}
			if ( adStr.find_last_of( "]")<adStr.length() )
			break ;
		}
		if (timeout==10)
			throw JobOperationException  ( __FILE__ , __LINE__ ,METHOD , WMS_JOBOP_ALLOWED , "Unable to read listener named pipe streams" );
	
		glite::jdl::Ad ad ( adStr );
		this->port = ad.getIntValue("PORT")[0];
		pid  = ad.getIntValue("PID")[0];
		if  (   fromP>0 &&   (   (this->port < fromP  ) || (this->port > toP)    )   ){
			detach();
			throw JobOperationException  ( __FILE__ , __LINE__ ,METHOD , WMS_JOBOP_ALLOWED , "Unable to perform attachement: port exceeds firewall range" );
		}
		GLITE_STACK_CATCH() ; //Exiting from method: remove line from stack trace
	};

} // api
} // wmsui
} // glite

File: org.glite.wms-ui.api-java/src/org/glite/wmsui/apij/Shadow.java
/*
* Shadow.java
*
* Copyright (c) 2001 The European DataGrid Project - IST programme, all rights reserved.
*
* Contributors are mentioned in the code there appropriate.
*/
package org.glite.wmsui.apij;
import  org.glite.jdl.JobAd ;
import java.net.InetAddress ;
import java.util.StringTokenizer ;
import java.net.UnknownHostException ; //InetAddress Exc
import  java.io.BufferedReader ; //Stream of info gathering from pipes
import  java.io.FileReader;
import  java.io.FileWriter;
import  java.io.BufferedWriter; //Stream of info gathering from pipes
import java.io.File ;
import condor.classad.* ;
import java.io.IOException ;



/**
 * This class provides the core management for interactive jobs.
 * once the glite-wms-grid-console-shadow has started successfully and the job is running
 * the user should interact with the submitted job  (or might have attached to a previous job)
* At the end of the interaction the background bypass process should be
 * killed and the I/O pipes have to be removed. This is done automatically by the 'detach' method.
 * The shadow class must be used togheter with an implementation of the Listener interface, which actually performs the
 * final visual interactivity with the user.
 * @see Listener
 *
 * @version 0.1
 * @author Alessandro Maraschini <alessandro.maraschini@datamat.it> */
public class Shadow implements Runnable{
	/**The constructor*/
	Shadow(JobId jid , Listener ls ){  listener = ls ;  jobId  = jid ;  };
	/**Class Destructor*/
	public void finalize () {
		// System.out.println ( "\n\n\nShadow::Finalising!!\n\n\n"  );
		try{     detach () ;  }
		catch (  Exception exc ){  
			/** Do nothing*/
		}
	};
	/**
	* Run the listener
	*/
	public void run() {
		if (listener!=null)  listener.run( this) ;
		else    finalize() ;
	};
	/**
	* Write into the input stream pipeline
	@param str the string to be written
	@throws throws IOException If any error occurred while writing in the pipe
	@throws NoSuchFieldException if the inputPipe has not been created yet
	*/
	public void writeIn (String str )throws IOException, NoSuchFieldException {
			if ( barbaWriter ==null ) barbaWriter = new BufferedWriter ( new FileWriter ( getPipeIn()  )  ) ;
			BufferedWriter inputBuffer = new BufferedWriter ( new FileWriter ( getPipeIn()  )  ) ;
			// System.out.println ("Shadow::writeln-->writing: " + str ) ;
			inputBuffer.write( str );
			inputBuffer.close ();
	}
	/**
	* Read any possible message stored inside the output pipeline
	* @return the message read in the pipe (if any)
	* @throws throws IOException If any error occurred while reading in the pipe
	* @throws NoSuchFieldException if the outputPipe has not been created yet */
	public String emptyOut ( )throws IOException, NoSuchFieldException {
		if ( outBuf ==null)  outBuf = new BufferedReader ( new FileReader(  getPipeOut() )  ) ;
		int intbuf = outBuf.read()  ;
		if (  intbuf!=-1  ) {
			char charbuf =  (char) intbuf ;
			String tmpStr =   String.valueOf (  charbuf  );
			return tmpStr ;
		}else { try{ Thread.sleep (1000) ; } catch (Exception exc) {} ;   return "" ; }
		// return outBuf.readLine() ;
	}

	/**
	Finalize the Shadow instance.
	Kill the background shadow process and delete the named pipes from the hard disc
	@throws java.lang.NoSuchFieldException if the pipes have not been created yet
	*/
	public void detach() throws java.lang.NoSuchFieldException{
		// Dleting the files:
		new File ( getPipeOut() ).delete();
		new File ( getPipeIn() ).delete();
		new File ( pipeRoot ).delete();
		// Killing the shadow proces:
		String command = "kill -9 "+ pid ;
		if (pid!=0)
			Api.shadow( command );
		pipeRoot=null ;
	};


	/**
	* Retrieve the name of the pipe that listens for the standard input
	* @return the string representation of the input pipe name for the current job
	* @throws java.lang.NoSuchFieldException the pipe has not been created yet
	*/
	public String getPipeIn() throws NoSuchFieldException {
		if (pipeRoot!=null) return pipeRoot +".in" ;
		else  throw new NoSuchFieldException("Shadow named pipe not yet created. Unable to retrieve")  ;
	};
	/**
	* Retrieve the name of the pipe that listens for the standard output
	* @return the string representation of the output pipe name for the current job
	@throws java.lang.NoSuchFieldException the pipe has not been created yet
	*/
	public String getPipeOut()throws NoSuchFieldException {
		if (pipeRoot!=null)
		return pipeRoot +".out" ;
		else  throw new NoSuchFieldException("Shadow named pipe not yet created. Unable to retrieve")  ;
	};

	/**
	* Retrieve the process id corresponding to the shadow background process
	* @return the id of the process related to the running shadow
	* @throws java.lang.NoSuchFieldException the process has not been launched yet */
	public int getPid() throws NoSuchFieldException , IOException  {
		if (pipeRoot!=null)
		return pid ;
		else  throw new NoSuchFieldException("Shadow named pipe not yet created. Unable to retrieve process Id")  ;
	}
	/**
	* Retrieve the Name of the host 
	* @return the string representation of the current host
	* @throws UnknownHostException the system is unable to retrieve host information */
	static public String getHost ()  throws UnknownHostException{
			String hostname = "" ;
			String ipAddr = "";
			InetAddress addr = InetAddress.getLocalHost();
			// System.out.println ( " addr = [" + addr.toString()+"]" );
			StringTokenizer tks = new StringTokenizer ( addr.toString() , "/" );
			if ( tks.countTokens() == 2 ){
				hostname = tks.nextToken ( );
				// ipAddr = tks.nextToken ( );
				return tks.nextToken ( );
			} else throw new UnknownHostException ("Unable to parse the Inet ip address") ;
	}
	/**
	* Retrieve the shadow port
	* @return the port where the shadow proces is listening to
	* @throws java.lang.NoSuchFieldException the process has not been launched yet */
	public int getPort()throws NoSuchFieldException  {
		if (pipeRoot!=null)   return port ;
		else  throw new NoSuchFieldException("Shadow named pipe not yet created. Unable to retrieve process Id")  ;
	}
	/**
	* Retrieve the id of the Job for the current Shadow
	* @return the JobId instance linked to the Shadow
	*/
	public JobId getJobId ()    {  return jobId ; }



	void start()   {  new Thread(  this ).start();  /* launch the listener */ };
	/**  Perform a grid-console-shadow  */
	int console( ) { return console (0 ); }
	/**  Perform a grid-console-shadow
	* @param port the port where the listener listen to from the agent   */
	int console( int port )   {
		pipeRoot = "/tmp/listener-" + jobId.unique;
		// Launch the grid-console-shadow with --logfile (background process)
		String shPath = Api.getEnv (GLITE_WMS_LOCATION) ;
		if (  shPath== null  )
			shPath = OPT_GLITE ;
		shPath = shPath +SEP +BIN +SEP + "glite-wms-grid-console-shadow"  ;
		File sh = new File ( shPath ) ;
		if (  !sh.isFile () )
			throw new RuntimeException ("Unable to find listener shadow executable: " +shPath ) ;
		String command = shPath ;
		String gtp_env  = Api.getEnv ("GLOBUS_TCP_PORT_RANGE") ;
		int fromP =0, toP =0;
		if (gtp_env!= null){
			// environment variable found
			String sep[] = {" " , ":" , "-" , "," } ;
			for (   int i = 0 ; i< sep.length ; i++  ){
				if ( gtp_env.indexOf( sep[i] ) !=-1  ){
					fromP =  java.lang.Integer.parseInt(   gtp_env.substring(0 , gtp_env.indexOf( sep[i] ) ).trim()  ) ;
					toP    =  java.lang.Integer.parseInt(   gtp_env.substring(gtp_env.indexOf( sep[i] )    ).trim()    );
					break;
				}
			}
		}
		if (port !=0){
			if (    (port < fromP  ) || (port > toP)    )
				throw new   RuntimeException( "Unable to perform attachement: port exceeds firewall range [" + gtp_env +"]");
			command += " -port " + port ;
		}
		command += " -log-to-file " + pipeRoot +"&";
		if ( Api.shadow( command  )  != 0 )
			throw new RuntimeException ("Unable to perform the listener shadow executable" ) ;
		/* Open the Pipe */
		int timeout = 0;
		BufferedReader pipe =null;
		boolean toSleep = false ;
		while ( timeout< 10 ){
			try {  if (toSleep){Thread.sleep ( 1000  ) ; toSleep=false ; } pipe = new BufferedReader ( new FileReader ( pipeRoot ) ) ; break ;
			}catch (Exception exc){  timeout++ ;  toSleep= true ; }
		}
		if (timeout ==10)  throw new RuntimeException ("Unable to open listener shadow info Stream:\n" + pipeRoot ) ;
		// get the pid and the port:
		String ad = new String();
		char[] buffer = new char[1];
		timeout=0 ;
		while (  timeout< 10  ){
			try {
			if (toSleep)
				{Thread.sleep ( 1000  ) ; toSleep=false ; }
			ad += String.valueOf (   (char) pipe.read()  ) ;
			if (  ad.trim().endsWith("]") )
				break ;
			}catch (  Exception exc){  timeout++ ;  toSleep= true ; }
		}
		if (timeout ==10)  throw new RuntimeException ("Unable to properly read the listener shadow info Stream:\n" + pipeRoot ) ;
		ClassAdParser cp = new ClassAdParser( ad  )  ;
		RecordExpr re = (RecordExpr) cp.parse() ;
		Constant co = (Constant) re.lookup("PORT") ;
		this.port = co.intValue() ;
		co = (Constant) re.lookup("PID") ;
		pid =co.intValue() ;
		if    (    fromP>0    &&   ( (this.port < fromP  ) || (this.port > toP))    ){
			// try { detach(); } catch (NoSuchFieldException exc) {}
			finalize();
			throw new   RuntimeException( "Unable to perform attachement: port exceeds firewall range [" + gtp_env +"]");
		}
		System.out.println ("Shadow::console-->Console Started Successffully") ;
		return 0 ;
	};
	/******
	*  private / packages Members:
	*******/
	int port ;	
	private int pid =0;
	private String host;
	private Listener listener ;
	private String pipeRoot = null;
	private BufferedReader outBuf , errBuf ;
	private BufferedWriter barbaWriter = null ;
	private JobId jobId;
	private static final String SEP = "/" ;
	private static final String GLITE_WMS_LOCATION = "GLITE_WMS_LOCATION";
	private static final String OPT_GLITE = SEP + "opt"+ SEP +  "glite";
	private static final String BIN = SEP + "bin" ;
}; //end Shaodw class

